File: tauri-app/src/types/live.ts
import type { LocalDataWarning } from "./diagnostics";
import type { CodexHomeSourceToken } from "./platform";
import type { ModelTokenBreakdown } from "./usage";

export interface CacheUsageAdvice {
  threadId: string;
  hitRate: number;
  affectedThreads?: number;
  low: boolean;
  timestamp: number;
}

export interface LiveRateSnapshot {
  cacheAdvice?: CacheUsageAdvice | null;
  scopeLabel: string;
  threadTitle: string;
  selectedThreadId: string | null;
  selectedThreadTitle: string;
  selectedTokensPerSecond: number;
  tokensPerSecond: number;
  totalTokens: number;
  totalTokensToday: number;
  requestsToday: number;
  maxTokensPerSecond: number;
  preciseEnabled: boolean;
  unreadSummary: UnreadSummary;
  warnings: LocalDataWarning[];
}

export interface LiveThreadOption {
  id: string;
  title: string;
  subtitle: string;
  updatedAt: string;
  tokensUsed: number;
}

export interface LiveRateStreamLease {
  leaseId: string;
  registered: boolean;
}

export interface FloatingPanelSnapshot {
  cacheAdvice?: CacheUsageAdvice | null;
  tokensPerSecond: number;
  maxTokensPerSecond: number;
  liveRateAvailable?: boolean;
  trendLabel: string;
  liveRateStatusKind?: "failure" | "pending";
  liveRateStatusLabel?: string;
  resetCreditLabel: string;
  resetCreditRateBarLabel?: string;
  resetCreditStandaloneLabel?: string;
  totalTokensLabel: string;
  todayTokensLabel: string;
  requestsLabel: string;
  todayModelBreakdowns: ModelTokenBreakdown[];
  fiveHourLabel: string;
  fiveHourAvailability: "measured" | "unavailable" | "absent";
  fiveHourRemainingPercent: number | null;
  /** Even-pace reference in percentage points (0–100), not the remaining ratio. */
  fiveHourExpectedRemainingPercent: number | null;
  sevenDayLabel: string;
  sevenDayAvailability: "measured" | "unavailable" | "absent";
  sevenDayRemainingPercent: number | null;
  /** Even-pace reference in percentage points (0–100), not the remaining ratio. */
  sevenDayExpectedRemainingPercent: number | null;
  /** The displayed quota is last-good data after a prolonged read failure. */
  quotaDataStale?: boolean;
  unread: boolean;
  unreadSummary: UnreadSummary;
}

export interface UnreadSummary {
  active: boolean;
  count: number;
  label: string;
  detail: string;
  source: string;
}

export interface UnreadSummaryChangedPayload {
  sourceToken: CodexHomeSourceToken;
  summary: UnreadSummary;
}

File: tauri-app/src-tauri/src/models/live.rs
use serde::Serialize;

use super::LocalDataWarning;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveRateSnapshot {
    pub scope_label: String,
    pub thread_title: String,
    pub selected_thread_id: Option<String>,
    pub selected_thread_title: String,
    pub selected_tokens_per_second: f64,
    pub tokens_per_second: f64,
    pub total_tokens_today: u64,
    pub requests_today: u32,
    pub max_tokens_per_second: f64,
    pub precise_enabled: bool,
    pub warnings: Vec<LocalDataWarning>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LiveThreadOption {
    pub id: String,
    pub title: String,
    pub subtitle: String,
    pub updated_at: String,
    pub tokens_used: u64,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FloatingPanelSnapshot {
    pub tokens_per_second: f64,
    pub trend_label: String,
    pub total_tokens_label: String,
    pub today_tokens_label: String,
    pub requests_label: String,
    pub five_hour_label: String,
    pub seven_day_label: String,
    pub unread: bool,
    pub unread_summary: UnreadSummary,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct UnreadSummary {
    pub active: bool,
    pub count: u32,
    pub label: String,
    pub detail: String,
    pub source: String,
}
